File: cs50/week2/problem-set/ceasar/caesar.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

void encrypt(string, int);
char increment(char);
char increment_lower(char);
char increment_upper(char);
bool is_number(string);

int main(int argc, string argv[]){
    if(argc != 2 || !is_number(argv[1])){
        printf("No key given or invalid.\n");
        return 1;
    }
    
    int k = atoi(argv[1]); // takes an integer from command line
    
    printf("plaintext: ");
    string secret = get_string();
    
    printf("ciphertext: ");
    encrypt(secret, k);;
}

void encrypt(string secret, int k){
    char letter = '\0';
    
    for(int i = 0, length = strlen(secret); i < length; i++){
        
        letter = secret[i];
        
        for(int j = 0; j < k; j++){
            letter = increment(letter);
        }
        
        printf("%c", letter);
    }
    
    printf("\n");
}

char increment(char letter){
    if(letter >= 'a' && letter <= 'z'){
        return increment_lower(letter);
    }
    
    if(letter >= 'A' && letter <= 'Z'){
        return increment_upper(letter);
    }
        
    return letter;
}

char increment_lower(char letter){
    if(letter == 'z'){
        return 'a';
    }
    
    return letter + 1; 
}

char increment_upper(char letter){
    if(letter == 'Z'){
        return 'A';
    }
    
    return letter + 1;
}

bool is_number(string number){
    for(int i = 0, length = strlen(number); i < length; i++){
        if(!('0' <= number[i] && '9' >= number[i])){
            return false;
        }
    }
    
    return true;
}

File: cs50/week8/problem-set/caesar.py
# Program that takes in a number as a command line argument and
# uses it as a guide for the rotation in a caesar cypher.
# Spec here: http://docs.cs50.net/problems/caesar/caesar.html
# Assignment to convert from C to python here:
# http://docs.cs50.net/2017/x/psets/6/pset6.html
# Code by Roman Davis
# Code honestly seems a little dumb and unpythonic, though it's a
# very faithful translation of my C solution.

import cs50
import sys

def main():
    if len(sys.argv) != 2:
        print("Must have one argument.")
        exit(1)
        
    print("plaintext: ", end = "")
    plaintext = cs50.get_string()
    
    if plaintext == None:
        print("get_string() failed.")
        exit(2)
        
    print("cyphertext: ", end = "")
    print(cypher(plaintext, int(sys.argv[1])))
    
def cypher(text, n):
    cypher = ""
    for char in text:
        cypher += rotate(char, n)
        
    return cypher
    
def rotate(char, n):
    result = char
    for i in range(n):
        result = increment(result)
    
    return result
        
def increment(char):
    if ord(char) >= ord("A") and ord(char) <= ord("Z"):
        return increment_upper(char)
    elif ord(char) >= ord("a") and ord(char) <= ord("z"):
         return increment_lower(char)
    else:
        return char
        
def increment_upper(char):
    if char == "Z":
        return "A"
    return chr(ord(char) + 1)
    
def increment_lower(char):
    if char == "z":
        return "a"
    return chr(ord(char) + 1)
    
if __name__ == "__main__":
    main()